src/main/java/com/puppycrawl/tools/checkstyle/site/ViolationMessagesMacro.java

```diff
@@ -0,0 +1,280 @@
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.site;
+
+import java.lang.reflect.Field;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.TreeSet;
+
+import org.apache.maven.doxia.macro.AbstractMacro;
+import org.apache.maven.doxia.macro.Macro;
+import org.apache.maven.doxia.macro.MacroExecutionException;
+import org.apache.maven.doxia.macro.MacroRequest;
+import org.apache.maven.doxia.module.xdoc.XdocSink;
+import org.apache.maven.doxia.sink.Sink;
+import org.codehaus.plexus.component.annotations.Component;
+
+import com.puppycrawl.tools.checkstyle.ModuleFactory;
+import com.puppycrawl.tools.checkstyle.PackageNamesLoader;
+import com.puppycrawl.tools.checkstyle.PackageObjectFactory;
+import com.puppycrawl.tools.checkstyle.api.CheckstyleException;
+import com.puppycrawl.tools.checkstyle.checks.coding.AbstractSuperCheck;
+import com.puppycrawl.tools.checkstyle.checks.naming.AbstractAccessControlNameCheck;
+import com.puppycrawl.tools.checkstyle.checks.naming.AbstractNameCheck;
+import com.puppycrawl.tools.checkstyle.checks.regexp.RegexpMultilineCheck;
+import com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineCheck;
+import com.puppycrawl.tools.checkstyle.checks.regexp.RegexpSinglelineJavaCheck;
+import com.puppycrawl.tools.checkstyle.checks.whitespace.AbstractParenPadCheck;
+
+/**
+ * A macro that inserts an unordered list of the violation messages.
+ */
+@Component(role = Macro.class, hint = "violation-messages")
+public class ViolationMessagesMacro extends AbstractMacro {
+    @Override
+    public void execute(Sink sink, MacroRequest request) throws MacroExecutionException {
+        final String moduleName = (String) request.getParameter("moduleName");
+        final Class<?> clss = getModuleClass(moduleName);
+        final Set<Field> messageKeyFields = getCheckMessageKeyFields(clss);
+        final Set<String> messageKeys = getMessageKeys(messageKeyFields);
+        createListOfMessages(sink, clss, messageKeys);
+    }
+
+    /**
+     * Constructs a URL to GitHub that searches for the message key.
+     *
+     * @param clss the class of the module.
+     * @param messageKey the message key.
+     * @return the URL to GitHub.
+     */
+    private static String constructMessageKeyUrl(Class<?> clss, String messageKey) {
+        return "https://github.com/search?q="
```

Comment:
Url was taken from https://github.com/checkstyle/checkstyle/blob/0c2ab5338db569b4699d5d20121d1bd508394eb9/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1427-L1433

Right now we don't want to escape this text because we convert from template to regular xdoc. Regular xdoc expects it to be unescaped and default xdoc parser and sink will take care of escaping when generating the website(`mvn site`):
https://github.com/apache/maven-doxia/blob/master/doxia-core/src/main/java/org/apache/maven/doxia/sink/impl/Xhtml5BaseSink.java#L1595
This was part of the reason why I overwrote `link(x)` in `XdocsTemplateSink`

Reply:
I did one final try with `UrlEncoder.encode` and it works. Almost. 
```java
URLEncoder.encode("....", StandardCharsets.UTF_8);
```
https://www.baeldung.com/java-url-encoding-decoding

A space can either be encoded as a `%20` or `+`. This function encodes it as a `+`. The problem is that the following test  expects both `%20` and `+`.

https://github.com/checkstyle/checkstyle/blob/0c2ab5338db569b4699d5d20121d1bd508394eb9/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1427-L1433

Shall we move this to an issue and deal with it after all violation messages are generated from the macro? Then I can do the encoding and change the test.

Reply:
Yes, let's make another issue.

Reply:
Opened at https://github.com/checkstyle/checkstyle/issues/13490